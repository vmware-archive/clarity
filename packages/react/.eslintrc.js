const OFF = 0;
const WARN = 1;
const ERROR = 2;

const bannedTSTypes = {
  Array: 'Use [] instead.',
  Object: 'Use {} instead.',
  Boolean: 'Use `boolean` instead.',
  Number: 'Use `number` instead.',
  String: 'Use `string` instead.',
};

// only adding rules that override the defaults or enforce new standards
const rules = {
  '@typescript-eslint/explicit-function-return-type': OFF, // too much work at the moment
  '@typescript-eslint/no-explicit-any': OFF, // we have custom events whose type is any; revisit if needed
  '@typescript-eslint/no-use-before-define': OFF,
  '@typescript-eslint/ban-types': [ERROR, { types: bannedTSTypes }],
  '@typescript-eslint/explicit-member-accessibility': [ERROR, { accessibility: 'no-public' }],
  '@typescript-eslint/no-unused-vars': [
    ERROR,
    {
      varsIgnorePattern: 'React|Cds',
    },
  ],
  'import/extensions': [ERROR, 'ignorePackages'],
};

const parserOptions = {
  project: 'tsconfig.eslint.json',
  tsconfigRootDir: __dirname,
  sourceType: 'module',
};

const overrides = [
  {
    files: ['*.test.tsx'],
    rules: {
      'import/extensions': OFF,
    },
  },
];

const config = {
  extends: '../../.eslintrc.js',
  parserOptions,
  rules,
  overrides,
};

module.exports = config;
