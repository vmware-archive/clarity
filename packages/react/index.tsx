import React from 'react';
import ReactDOM from 'react-dom';
import 'normalize.css/normalize.css';
import '@cds/core/global.min.css';
import '@cds/city/css/bundles/default.min.css';
import App from './App';

ReactDOM.render(<App />, document.getElementById('root'));
