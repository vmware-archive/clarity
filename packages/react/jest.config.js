module.exports = {
  roots: ['./'],
  globals: {
    'ts-jest': {
      tsConfig: 'tsconfig.test.json',
    },
  },
  modulePathIgnorePatterns: ['<rootDir>/dist/'],
  moduleNameMapper: {
    '@cds/(.*)': '<rootDir>/../core/dist/$1',
  },
  setupFilesAfterEnv: ['./jest.setup.ts'],
  snapshotSerializers: ['enzyme-to-json/serializer'],
  testMatch: ['**/__tests__/**/*.+(ts|tsx|js)', '**/?(*.)+(test).+(ts|tsx|js)'],
  transform: {
    '^.+\\.(js|jsx)$': 'babel-jest',
    '^.+\\.(ts|tsx)$': 'ts-jest',
  },
  transformIgnorePatterns: ['node_modules/(?!@cds/core|lit-element|lit-html|ramda)'],
};
