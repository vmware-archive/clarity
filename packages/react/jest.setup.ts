import '@testing-library/jest-dom/extend-expect';

if (typeof window !== 'undefined') {
  // jsdom doesn't implment IntersectionObserver
  // https://github.com/jsdom/jsdom/issues/2032
  const mockObserverAPI = () =>
    jest.fn().mockReturnValue({
      observe: jest.fn(),
      unobserve: jest.fn(),
      disconnect: jest.fn(),
    });

  window.ResizeObserver = mockObserverAPI();
  window.IntersectionObserver = mockObserverAPI();

  // jsdom implements getBoundingClientRect, but not DOMRect
  if (!window.DOMRect) {
    const domRectMock = () => jest.fn().mockReturnValue({});
    window.DOMRect = (domRectMock() as unknown) as typeof window.DOMRect;
  }
}
