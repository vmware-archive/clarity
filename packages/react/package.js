'use strict';

const fs = require('fs-extra');
const path = require('path');
const del = require('del');

// Temporary script needed to remove cache files and
// files Angular CLI leaves in referenced dependencies
// https://github.com/ng-packagr/ng-packagr/pull/1372
// https://github.com/ng-packagr/ng-packagr/issues/1318
// https://github.com/angular/angular/issues/33395
// https://justinfagnani.com/2019/11/01/how-to-publish-web-components-to-npm/
const read = dir =>
  fs
    .readdirSync(dir)
    .reduce(
      (files, file) =>
        fs.statSync(path.join(dir, file)).isDirectory()
          ? files.concat(read(path.join(dir, file)))
          : files.concat(path.join(dir, file)),
      []
    );

del.sync(
  [
    './dist/react/**/*.{tsbuildinfo,test.js,test.js.map,test.d.ts,.snap}',
    './dist/react/*.{tsbuildinfo,test.js,test.js.map,test.d.ts,.snap}',
  ],
  { force: true }
);

read('./dist/react')
  .filter(f => f.includes('package.json'))
  .forEach(file => {
    const packageFile = fs.readJsonSync(file);
    const metaData = {
      main: './index.js',
      module: './index.js',
      typings: './index.d.ts',
      type: 'module',
    };

    if (file === 'dist/react/package.json') {
      ['alias', 'browserslist', 'scripts', 'devDependencies'].forEach(p => delete packageFile[p]);
      // move @clr/core from dependencies to peerDependencies and assign version number that matches the package.json
      packageFile.peerDependencies = { '@clr/core': `^${packageFile.version}` };
      delete packageFile.dependencies['@clr/core'];
    }

    fs.writeJsonSync(file, { ...packageFile, ...metaData }, { spaces: 2 });
  });
