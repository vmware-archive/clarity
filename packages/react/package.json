{
  "name": "@cds/react",
  "version": "5.0.0-next.0",
  "scripts": {
    "start": "npm-run-all core:build dev:start",
    "dev:start": "yarn run clean && parcel index.html --cache-dir=./.parcel-cache",
    "build": "npm-run-all -s build:ts build:package build:update",
    "build:preinstall": "mkdir -p ../../dist/react/ && cp ./package.json ../../dist/react/",
    "build:update": "node ../../scripts/update-local-packages.js /packages/react/dist/react/package.json",
    "build:package": "node ./package.js",
    "build:ts": "tsc --b ./tsconfig.project.json --force",
    "core:build": "yarn --cwd ../core run build",
    "lint": "eslint \"*/**/*.{ts,tsx}\" --ignore-pattern \"dist\"",
    "lint:fix": "eslint --fix \"*/**/*.{ts,tsx}\" --ignore-pattern \"dist\"",
    "test": "jest --coverage",
    "test:watch": "jest --watchAll",
    "test:snap:update": "jest --updateSnapshot",
    "clean": "del ./dist .parcel-cache"
  },
  "dependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1"
  },
  "peerDependencies": {
    "@cds/core": "./dist/core"
  },
  "devDependencies": {
    "@babel/core": "7.10.2",
    "@babel/preset-env": "7.10.2",
    "@types/enzyme": "^3.10.5",
    "@types/enzyme-adapter-react-16": "^1.0.6",
    "@types/jest": "^26.0.0",
    "@types/node": "^14.0.5",
    "@types/react": "^16.9.35",
    "@types/react-dom": "^16.9.8",
    "@types/react-test-renderer": "^16.9.2",
    "@typescript-eslint/eslint-plugin": "2.27.0",
    "@typescript-eslint/parser": "2.27.0",
    "babel-jest": "26.0.1",
    "cpy-cli": "3.1.0",
    "del-cli": "1.1.0",
    "enzyme": "^3.11.0",
    "enzyme-adapter-react-16": "^1.15.2",
    "enzyme-to-json": "^3.5.0",
    "eslint": "6.8.0",
    "eslint-config-prettier": "6.10.1",
    "jest": "^26.0.1",
    "npm-run-all": "4.1.5",
    "parcel": "^2.0.0-beta.1",
    "prettier": "2.0.5",
    "pretty-quick": "1.11.0",
    "replace": "1.1.0",
    "ts-jest": "^26.1.0",
    "typescript": "^4.0.0"
  },
  "optionalDependencies": {
    "@cds/city": "^1.1.0",
    "normalize.css": "^8.0.1"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "author": "clarity",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/vmware/clarity/issues"
  }
}