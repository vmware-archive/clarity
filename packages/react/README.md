# Clarity React Wrappers for Core Web Components [![npm version](https://badge.fury.io/js/%40clr%2Freact.svg)](https://badge.fury.io/js/%40clr%2Freact)

Clarity Core is a suite of Web Components from the [Clarity Design System](https://clarity.design). Because React [doesn't fully interoperate with custom elements](https://custom-elements-everywhere.com/) we have developed this library of React components that wrap Clarity Web Components

## Installation

1.  (Optional): Install the Clarity Core package from npm.

    ```bash
    npm install @cds/core --save
    ```

2.  Install the Clarity React package from npm.

    ```bash
    npm install @cds/react --save
    ```

3.  Import desired component into your JavaScript or TypeScript

    ```typescript
    import { CdsModal, CdsModalActions, CdsModalContent, CdsModalHeader } from '@cds/react/modal';
    ```

4.  Use React wrapped Web Component in React

    ```jsx
    {
      /*
      Example of a modal web component in React with React Shim
      - size - attribute style hook
      - open - setting the 'open' property on the element
      - openChange - listen for the `openChange` custom event
    */
    }
    <CdsModal onCloseChange={this.handleCloseModalClick.bind(this, false)}>
      <CdsModalHeader>
        <h3 cds-text="title">My Modal</h3>
      </CdsModalHeader>
      <CdsModalContent>
        <div cds-layout="vertical gap:md p-y:xs">
          <p cds-text="body">Lorem Ipsum</p>
        </div>
      </CdsModalContent>
      <CdsModalActions>
        <div cds-layout="horizontal gap:sm align:right">
          <CdsButton onClick={this.handleCloseModalClick.bind(this, false)} action="outline">
            Cancel
          </CdsButton>
          <CdsButton onClick={this.handleCloseModalClick.bind(this, false)}>Ok</CdsButton>
        </div>
      </CdsModalActions>
    </CdsModal>;
    ```

## Using refs

In React [refs](https://reactjs.org/docs/refs-and-the-dom.html) provide a way to access DOM nodes or
React elements created in the render method. Because web components' lifecycle lives outside of react's
lifecycle our components provide a way to use refs when the underlying web component has finished rendering:

```typescript
import React from 'react';
import { CdsButton } from '@cds/react/button';

export default class App extends React.Component<{}, {}> {
  buttonRef: React.RefObject<CdsButton>;

  constructor(props: any) {
    super(props);
    this.buttonRef = React.createRef<CdsButton>();
  }

  componentDidMount() {
    this.buttonRef.current.nativeElement.then(element => {
      element.focus();
    });
  }

  render() {
    return (
      <div>
        <CdsButton ref={this.buttonRef}>My button</CdsButton>
      </div>
    );
  }
}
```
