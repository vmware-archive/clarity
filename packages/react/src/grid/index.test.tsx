import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { CdsGrid, CdsGridColumn, CdsGridRow, CdsGridCell, CdsGridFooter } from './index';

describe('CdsGrid', () => {
  it('renders', () => {
    const wrapper = shallow(
      <div>
        <CdsGrid aria-label="basic datagrid demo">
          <CdsGridColumn>Type</CdsGridColumn>
          <CdsGridColumn>Description</CdsGridColumn>
          <CdsGridColumn>Amount</CdsGridColumn>
          <CdsGridColumn>Balance</CdsGridColumn>
          <CdsGridRow>
            <CdsGridCell>Deposit</CdsGridCell>
            <CdsGridCell>Item</CdsGridCell>
            <CdsGridCell>$1,000,000.00</CdsGridCell>
            <CdsGridCell>$1,000,000.00</CdsGridCell>
          </CdsGridRow>
          <CdsGridRow>
            <CdsGridCell>Credit</CdsGridCell>
            <CdsGridCell>Billing</CdsGridCell>
            <CdsGridCell>$250.00</CdsGridCell>
            <CdsGridCell>$523,750.00</CdsGridCell>
          </CdsGridRow>
          <CdsGridFooter>Footer</CdsGridFooter>
        </CdsGrid>
      </div>
    );
    const renderedComponent = wrapper.find('CdsGrid');
    expect(renderedComponent.html()).toMatch(/Type/);
    expect(renderedComponent.html()).toMatch(/Deposit/);
    expect(renderedComponent.html()).toMatch(/Footer/);
  });

  it('snapshot', () => {
    const wrapper = mount(
      <div>
        <CdsGrid aria-label="basic datagrid demo">
          <CdsGridColumn>Type</CdsGridColumn>
          <CdsGridColumn>Description</CdsGridColumn>
          <CdsGridColumn>Amount</CdsGridColumn>
          <CdsGridColumn>Balance</CdsGridColumn>
          <CdsGridRow>
            <CdsGridCell>Deposit</CdsGridCell>
            <CdsGridCell>Item</CdsGridCell>
            <CdsGridCell>$1,000,000.00</CdsGridCell>
            <CdsGridCell>$1,000,000.00</CdsGridCell>
          </CdsGridRow>
          <CdsGridRow>
            <CdsGridCell>Credit</CdsGridCell>
            <CdsGridCell>Billing</CdsGridCell>
            <CdsGridCell>$250.00</CdsGridCell>
            <CdsGridCell>$523,750.00</CdsGridCell>
          </CdsGridRow>
          <CdsGridFooter></CdsGridFooter>
        </CdsGrid>
      </div>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
