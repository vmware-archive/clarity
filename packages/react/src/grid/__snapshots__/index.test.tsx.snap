// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdsGrid snapshot 1`] = `
<div>
  <CdsGrid
    aria-label="basic datagrid demo"
  >
    <cds-grid(CustomElement)
      __forwardedRef={null}
      aria-label="basic datagrid demo"
    >
      <cds-grid
        __forwardedRef={null}
        aria-label="basic datagrid demo"
      >
        <CdsGridColumn>
          <cds-grid-column(CustomElement)
            __forwardedRef={null}
          >
            <cds-grid-column
              __forwardedRef={null}
            >
              Type
            </cds-grid-column>
          </cds-grid-column(CustomElement)>
        </CdsGridColumn>
        <CdsGridColumn>
          <cds-grid-column(CustomElement)
            __forwardedRef={null}
          >
            <cds-grid-column
              __forwardedRef={null}
            >
              Description
            </cds-grid-column>
          </cds-grid-column(CustomElement)>
        </CdsGridColumn>
        <CdsGridColumn>
          <cds-grid-column(CustomElement)
            __forwardedRef={null}
          >
            <cds-grid-column
              __forwardedRef={null}
            >
              Amount
            </cds-grid-column>
          </cds-grid-column(CustomElement)>
        </CdsGridColumn>
        <CdsGridColumn>
          <cds-grid-column(CustomElement)
            __forwardedRef={null}
          >
            <cds-grid-column
              __forwardedRef={null}
            >
              Balance
            </cds-grid-column>
          </cds-grid-column(CustomElement)>
        </CdsGridColumn>
        <CdsGridRow>
          <cds-grid-row(CustomElement)
            __forwardedRef={null}
          >
            <cds-grid-row
              __forwardedRef={null}
            >
              <CdsGridCell>
                <cds-grid-cell(CustomElement)
                  __forwardedRef={null}
                >
                  <cds-grid-cell
                    __forwardedRef={null}
                  >
                    Deposit
                  </cds-grid-cell>
                </cds-grid-cell(CustomElement)>
              </CdsGridCell>
              <CdsGridCell>
                <cds-grid-cell(CustomElement)
                  __forwardedRef={null}
                >
                  <cds-grid-cell
                    __forwardedRef={null}
                  >
                    Item
                  </cds-grid-cell>
                </cds-grid-cell(CustomElement)>
              </CdsGridCell>
              <CdsGridCell>
                <cds-grid-cell(CustomElement)
                  __forwardedRef={null}
                >
                  <cds-grid-cell
                    __forwardedRef={null}
                  >
                    $1,000,000.00
                  </cds-grid-cell>
                </cds-grid-cell(CustomElement)>
              </CdsGridCell>
              <CdsGridCell>
                <cds-grid-cell(CustomElement)
                  __forwardedRef={null}
                >
                  <cds-grid-cell
                    __forwardedRef={null}
                  >
                    $1,000,000.00
                  </cds-grid-cell>
                </cds-grid-cell(CustomElement)>
              </CdsGridCell>
            </cds-grid-row>
          </cds-grid-row(CustomElement)>
        </CdsGridRow>
        <CdsGridRow>
          <cds-grid-row(CustomElement)
            __forwardedRef={null}
          >
            <cds-grid-row
              __forwardedRef={null}
            >
              <CdsGridCell>
                <cds-grid-cell(CustomElement)
                  __forwardedRef={null}
                >
                  <cds-grid-cell
                    __forwardedRef={null}
                  >
                    Credit
                  </cds-grid-cell>
                </cds-grid-cell(CustomElement)>
              </CdsGridCell>
              <CdsGridCell>
                <cds-grid-cell(CustomElement)
                  __forwardedRef={null}
                >
                  <cds-grid-cell
                    __forwardedRef={null}
                  >
                    Billing
                  </cds-grid-cell>
                </cds-grid-cell(CustomElement)>
              </CdsGridCell>
              <CdsGridCell>
                <cds-grid-cell(CustomElement)
                  __forwardedRef={null}
                >
                  <cds-grid-cell
                    __forwardedRef={null}
                  >
                    $250.00
                  </cds-grid-cell>
                </cds-grid-cell(CustomElement)>
              </CdsGridCell>
              <CdsGridCell>
                <cds-grid-cell(CustomElement)
                  __forwardedRef={null}
                >
                  <cds-grid-cell
                    __forwardedRef={null}
                  >
                    $523,750.00
                  </cds-grid-cell>
                </cds-grid-cell(CustomElement)>
              </CdsGridCell>
            </cds-grid-row>
          </cds-grid-row(CustomElement)>
        </CdsGridRow>
        <CdsGridFooter>
          <cds-grid-footer(CustomElement)
            __forwardedRef={null}
          >
            <cds-grid-footer
              __forwardedRef={null}
            />
          </cds-grid-footer(CustomElement)>
        </CdsGridFooter>
      </cds-grid>
    </cds-grid(CustomElement)>
  </CdsGrid>
</div>
`;
